javalib/src/main/scala/java/util/regex/Pattern.scala

```diff
@@ -118,18 +41,157 @@ object Pattern {
       res
     }
 
-  def adaptPatternToRe2(regex: String): String = {
-    regex
+  private object CompiledPatternStore {
+    private case class Entry(regex: String, flags: Int, re2: RE2RegExpOps)
+
+    private val cache =
+      scala.collection.mutable.IndexedSeq.fill[Entry](100)(null)
```

Comment:
Given that the memory for compiled regular expressions is never going to be freed, the overhead of storing them permanently is negligible. Lets use normal (non-fixed-size) hash map for the cache instead.

Reply:
Compiled regexes can be and actually are freed in`put` (l. 76). That said, I agree with you that a normal map is appropriate. How about using (Scala's) `LinkedHashMap` with a separate `size` field? This way, we can keep cache with a Map while limiting the memory usage by keeping `tail`s.

Reply:
Lets try the following then:

```
final case class Key(regex: String, flags: Int)
final class Node(var key: Key, var value: RE2RegExpOps, var rc: Int, var next: Node)

var current: Node = ???
var map: java.util.HashMap[Key, Node] = ???
```

1. Nodes form a ringbuffer of some fixed length (lets say 128).
   Every node stores a key that identifies the value it contains,
   precomplied regular expression and a reference count of current
   users of given regular expression.

2. Apart from the ringbuffer there is also a map from keys to 
   nodes that provides direct mapping to already precompiled regular expressions.

3. Whenever a key is requested we do the following:

   1. Check if it's in the map, if it is, lookup the regex through the map
      and increase the reference count for the node while it's being used. 

   2. If it's not, start traversing the ring buffer until a node with rc=0
      is found. We dispose of the old regular exression and remove the key
      from the map. Now the node can be re-used to store a new key/value pair.

   3. If while traversing we get back to the location where we started, 
      we need to grow a ringbuffer by 1 more node, and then we can do the same as ii.

Reply:
You realize that your algorithm never decrements `rc`? Obviously it can't, since you don't know when someone does not use a regex anymore.

Just use an LRU and stop worrying.

Reply:
@sjrd The only access to the store is through withRE2Regex so it does know when to decrement and increment.

Reply:
It also scales trivially to multithreaded equivalent with CAS and atomic RC in presence of different threads using different regexes at their own pace.

Reply:
@densh Implemented your suggestion.